packages/block-editor/src/components/block-variation-transforms/index.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * External dependencies
+ */
+import { isMatch } from 'lodash';
+
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import {
+	DropdownMenu,
+	MenuGroup,
+	MenuItemsChoice,
+} from '@wordpress/components';
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useState, useEffect } from '@wordpress/element';
+import { chevronDown } from '@wordpress/icons';
+
+const getMatchedVariation = ( blockAttributes, variations ) => {
```

Comment:
It looks quite complex so having a unit test covering it would help optimize the internal implementation.

`find` is one way of approaching it, although it returns the first matching element so in the very unlikely case  where two variations set the same attributes it would always pick the first one. I don't know if we should really care about it, but I included an alternative that would use filter:
```js
if ( ! variations?.length ) return; // or it could be removed by defaulting to an empty array in func param
const matches = variations.filter(
    ( { attributes = {} } ) => {
       const keys =  Object.keys(  attributes );
       return keys.length > 0 && isEqual( attributes, pick( blockAttributes, keys ) );
    }
);
if ( matches.length !== 1 ) return null;
return matches[0];
```

Reply:
>I don't know if we should really care about it

I think that we shouldn't for now.. :) But if you think it's better I can change it to `filter`. Although don't you think `isMatch` of lodash would be good to be used in `filter`? `isEqual` and `pick` implementation do something differently? 

I'm writing unit tests now.

Reply:
It isn't important whether you use `isMatch` or `isEqual`, I wrote this code to present the idea.

Reply:
Actually during my tests I realized that `filter` it's better than `find`, even if this is a simple matching function. Thanks for the input @gziolo ! 👍